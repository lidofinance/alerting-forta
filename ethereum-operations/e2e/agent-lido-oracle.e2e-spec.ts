import { configureContainer, Finding } from "forta-agent";
import { AwilixContainer, asFunction } from "awilix";
import {
  provideAgentPath,
  provideRunBlock,
  provideRunTransaction,
} from "./utils";

const TEST_TIMEOUT = 60_000; // ms

describe("agent-lido-oracle e2e tests", () => {
  let runBlock: (blockHashOrNumber: string | number) => Promise<Finding[]>;
  let runTransaction: (txHash: string) => Promise<Finding[]>;
  let logSpy: jest.SpyInstance;

  beforeAll(() => {
    logSpy = jest.spyOn(console, "log");
    logSpy.mockImplementation(() => {});
  });

  beforeEach(async () => {
    const container = configureContainer() as AwilixContainer;
    container.register({
      agentPath: asFunction(
        provideAgentPath("subagents/lido-oracle/agent-lido-oracle")
      ),
      runTransaction: asFunction(provideRunTransaction),
      runBlock: asFunction(provideRunBlock),
    });

    // https://docs.forta.network/en/latest/cli/#invoke-commands-programmatically
    runTransaction = container.resolve("runTransaction");
    runBlock = container.resolve("runBlock");
  });

  afterAll(() => {
    jest.resetAllMocks();
  });

  it(
    "should process block with Lido Oracle report overdue",
    async () => {
      const findings = await runBlock(15953817);
      expect(findings.at(0)).toMatchSnapshot();
    },
    TEST_TIMEOUT
  );

  it(
    "should process block with low balance of Lido Oracle",
    async () => {
      const findings = await runBlock(16191626);
      expect(findings.at(0)).toMatchSnapshot();
    },
    TEST_TIMEOUT
  );

  it(
    "should process tx with Lido Oracle report",
    async () => {
      const findings = await runTransaction(
        "0xe949652989ceed222ad1d1a903f7c925d64a7227b6a286451f9a454f753e9241"
      );
      expect(findings.at(0)).toMatchSnapshot();
    },
    TEST_TIMEOUT
  );

  it(
    "should process tx with decreased Lido Beacon rewards",
    async () => {
      const findings = await runTransaction(
        "0x1a5eed94c2da9da1ab5d40b723f92c43ce8a06e00b0a369d15561618115ef199"
      );
      expect(findings.at(1)).toMatchSnapshot();
    },
    TEST_TIMEOUT
  );

  it(
    "should process tx with super sloppy Lido Oracle",
    async () => {
      const findings = await runTransaction(
        "0xc77f7ba9780b2bff72f8bd7cfe16043342644bec7fec42561785a2aac0729cf2"
      );
      expect(findings.at(1)).toMatchSnapshot();
    },
    TEST_TIMEOUT
  );

  it(
    "should process tx with changed allowed Beacon Balance Annual Relative Increase",
    async () => {
      const findings = await runTransaction(
        "0xf3b49e6ebf2db6ae64fb96f1ea03e32233fd0cb7219f9256907496fc76028e39"
      );
      expect(findings.at(0)).toMatchSnapshot();
    },
    TEST_TIMEOUT
  );

  it(
    "should process tx with changed allowed Beacon Balance Annual Relative Decrease",
    async () => {
      const findings = await runTransaction(
        "0x9a1de3b55a77ce539d092b3a4b36bdd4ff179aafc03c723235c79aa21c5670f0"
      );
      expect(findings.at(1)).toMatchSnapshot();
    },
    TEST_TIMEOUT
  );

  it(
    "should process tx with changed Beacon Report Receiver",
    async () => {
      const findings = await runTransaction(
        "0x3f126198f641194e96cf6be0db03efc6ae92a4e1586b5062aaf373f171f0c5d2"
      );
      expect(findings.at(0)).toMatchSnapshot();
    },
    TEST_TIMEOUT
  );

  it(
    "should process tx with added member",
    async () => {
      const findings = await runTransaction(
        "0x4754b74262afb0c0b0c344a1d0600857e23634fdf43ba32d7752cc0cec1626b1"
      );
      expect(findings.at(0)).toMatchSnapshot();
    },
    TEST_TIMEOUT
  );

  it(
    "should process tx with changed quorum",
    async () => {
      const findings = await runTransaction(
        "0xd09be189f24903c8dc58676c22bd6b9ffc39bbb3f9f5c5a8613e3d198e310159"
      );
      expect(findings.at(4)).toMatchSnapshot();
    },
    TEST_TIMEOUT
  );
});
