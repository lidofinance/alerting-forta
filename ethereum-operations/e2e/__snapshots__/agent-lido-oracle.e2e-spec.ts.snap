// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`agent-lido-oracle e2e tests should process tx with Lido Oracle report 1`] = `
Finding {
  "addresses": Array [],
  "alertId": "LIDO-ORACLE-REPORT",
  "description": "Total balance: 5909524.783 ETH, total validators: 176728, rewards: 589.403 ETH (+0.609 ETH), time since last report: 23 hrs 54 min 24 sec",
  "labels": Array [],
  "metadata": Object {
    "beaconBalance": "5909524783080576000000000",
    "beaconValidators": "176728",
    "reportDelay": "86064",
    "rewards": "589402649547000000000",
    "rewardsDiff": "608666205000000000",
    "rewardsDiffPercent": "0.1",
  },
  "name": "‚úÖ Lido Oracle report",
  "protocol": "ethereum",
  "severity": 1,
  "type": 4,
}
`;

exports[`agent-lido-oracle e2e tests should process block with Lido Oracle report overdue 1`] = `
Finding {
  "addresses": Array [],
  "alertId": "LIDO-ORACLE-OVERDUE",
  "description": "Time since last report: 47 hrs 36 min 48 sec",
  "labels": Array [],
  "metadata": Object {
    "delay": "171408",
  },
  "name": "üö® Lido Oracle report overdue",
  "protocol": "ethereum",
  "severity": 5,
  "type": 3,
}
`;

exports[`agent-lido-oracle e2e tests should process tx with added member 1`] = `
Finding {
  "addresses": Array [],
  "alertId": "LIDO-ORACLE-MEMBER-ADDED",
  "description": "New oracle member added - 0x140Bd8FbDc884f48dA7cb1c09bE8A2fAdfea776E",
  "labels": Array [],
  "metadata": Object {
    "args": "0x140Bd8FbDc884f48dA7cb1c09bE8A2fAdfea776E",
  },
  "name": "‚ÑπÔ∏è Lido Oracle: Member Added",
  "protocol": "ethereum",
  "severity": 4,
  "type": 4,
}
`;

exports[`agent-lido-oracle e2e tests should process tx with changed Beacon Report Receiver 1`] = `
Finding {
  "addresses": Array [],
  "alertId": "LIDO-ORACLE-BEACON-REPORT-RECEIVER-SET",
  "description": "New beacon report receiver was set to 0x55a7E1cbD678d9EbD50c7d69Dc75203B0dBdD431",
  "labels": Array [],
  "metadata": Object {
    "args": "0x55a7E1cbD678d9EbD50c7d69Dc75203B0dBdD431",
  },
  "name": "‚ö†Ô∏è Lido Oracle: Beacon Report Receiver Change",
  "protocol": "ethereum",
  "severity": 4,
  "type": 4,
}
`;

exports[`agent-lido-oracle e2e tests should process tx with changed allowed Beacon Balance Annual Relative Decrease 1`] = `
Finding {
  "addresses": Array [],
  "alertId": "LIDO-ORACLE-BALANCE-RELATIVE-DECREASE-SET",
  "description": "Allowed beacon balance annual relative decrease was set to 500",
  "labels": Array [],
  "metadata": Object {
    "args": "500",
  },
  "name": "‚ö†Ô∏è Lido Oracle: Allowed Beacon Balance Annual Relative Decrease Change",
  "protocol": "ethereum",
  "severity": 4,
  "type": 4,
}
`;

exports[`agent-lido-oracle e2e tests should process tx with changed allowed Beacon Balance Annual Relative Increase 1`] = `
Finding {
  "addresses": Array [],
  "alertId": "LIDO-ORACLE-BALANCE-RELATIVE-INCREASE-SET",
  "description": "Allowed beacon balance annual relative increase was set to 1750",
  "labels": Array [],
  "metadata": Object {
    "args": "1750",
  },
  "name": "‚ö†Ô∏è Lido Oracle: Allowed Beacon Balance Annual Relative Increase Change",
  "protocol": "ethereum",
  "severity": 4,
  "type": 4,
}
`;

exports[`agent-lido-oracle e2e tests should process tx with changed quorum 1`] = `
Finding {
  "addresses": Array [],
  "alertId": "LIDO-ORACLE-QUORUM-CHANGED",
  "description": "Quorum size was set to 5",
  "labels": Array [],
  "metadata": Object {
    "args": "5",
  },
  "name": "üö® Lido Oracle: Quorum Changed",
  "protocol": "ethereum",
  "severity": 4,
  "type": 4,
}
`;

exports[`agent-lido-oracle e2e tests should process tx with decreased Lido Beacon rewards 1`] = `
Finding {
  "addresses": Array [],
  "alertId": "LIDO-ORACLE-REWARDS-DECREASED",
  "description": "Rewards decreased from 594.775 ETH to 583.402 by -11.373 ETH (-1.91%)",
  "labels": Array [],
  "metadata": Object {
    "beaconBalance": "5899179792328567000000000",
    "beaconValidators": "176460",
    "prevRewards": "594774737778000000000",
    "reportDelay": "86412",
    "rewards": "583402230417000000000",
    "rewardsDiff": "-11372507361000000000",
    "rewardsDiffPercent": "-1.91",
  },
  "name": "üö® Lido Beacon rewards decreased",
  "protocol": "ethereum",
  "severity": 3,
  "type": 3,
}
`;

exports[`agent-lido-oracle e2e tests should process block with low balance of Lido Oracle 1`] = `
Finding {
  "addresses": Array [],
  "alertId": "LIDO-ORACLE-LOW-BALANCE",
  "description": "Balance of 0x946D3b081ed19173dC83Cd974fC69e1e760B7d78 (Stakefish) is 0.2479 ETH. This is rather low!",
  "labels": Array [],
  "metadata": Object {
    "balance": "0.247862040202533669",
    "oracle": "0x946D3b081ed19173dC83Cd974fC69e1e760B7d78",
  },
  "name": "‚ö†Ô∏è Low balance of Lido Oracle",
  "protocol": "ethereum",
  "severity": 1,
  "type": 3,
}
`;

exports[`agent-lido-oracle e2e tests should process tx with super sloppy Lido Oracle 1`] = `
Finding {
  "addresses": Array [],
  "alertId": "SLOPPY-LIDO-ORACLE",
  "description": "Oracle 0x140Bd8FbDc884f48dA7cb1c09bE8A2fAdfea776E (Chorus One) has not reported before the quorum for more than 2 weeks or have never reported yet",
  "labels": Array [],
  "metadata": Object {},
  "name": "‚ö†Ô∏è Super sloppy Lido Oracle",
  "protocol": "ethereum",
  "severity": 3,
  "type": 2,
}
`;
